File: .github/scripts/build-image-matrix.mjs
#!/usr/bin/env node

import fs from "node:fs";
import path from "node:path";

const manifests = process.argv.slice(2);
if (manifests.length === 0) {
  throw new Error("at least one image manifest is required");
}

const requireString = (value, field, manifestPath) => {
  if (typeof value !== "string" || value.trim() === "") {
    throw new Error(`${manifestPath}: ${field} must be a non-empty string`);
  }
  return value;
};

const entries = manifests.flatMap((manifestPath) => {
  const manifest = JSON.parse(fs.readFileSync(manifestPath, "utf8"));
  const component = requireString(
    manifest.component ?? path.basename(path.dirname(manifestPath)),
    "component",
    manifestPath,
  );

  if (!Array.isArray(manifest.images) || manifest.images.length === 0) {
    throw new Error(`${manifestPath}: images must be a non-empty array`);
  }

  return manifest.images.map((image, index) => {
    if (image === null || typeof image !== "object" || Array.isArray(image)) {
      throw new Error(`${manifestPath}: images[${index}] must be an object`);
    }

    return {
      component,
      name: requireString(image.name, `images[${index}].name`, manifestPath),
      source: requireString(image.source, `images[${index}].source`, manifestPath),
      destination: requireString(image.destination, `images[${index}].destination`, manifestPath),
    };
  });
});

const destinations = new Set();
for (const entry of entries) {
  if (destinations.has(entry.destination)) {
    throw new Error(`duplicate destination image: ${entry.destination}`);
  }
  destinations.add(entry.destination);
}

process.stdout.write(JSON.stringify({ include: entries }));

File: .github/scripts/build-image-matrix.rb
#!/usr/bin/env ruby

require "json"
require "yaml"

def require_string(value, field, manifest_path)
  unless value.is_a?(String) && !value.strip.empty?
    raise "#{manifest_path}: #{field} must be a non-empty string"
  end

  value
end

def load_manifest(manifest_path)
  YAML.safe_load(
    File.read(manifest_path, encoding: "UTF-8"),
    permitted_classes: [],
    permitted_symbols: [],
    aliases: false,
  )
rescue Psych::Exception => error
  raise "#{manifest_path}: #{error.message}"
end

begin
  manifests = ARGV
  raise "at least one image manifest is required" if manifests.empty?

  entries = manifests.flat_map do |manifest_path|
    manifest = load_manifest(manifest_path)
    raise "#{manifest_path}: manifest must be a mapping" unless manifest.is_a?(Hash)

    component = require_string(
      manifest["component"] || File.basename(File.dirname(manifest_path)),
      "component",
      manifest_path,
    )
    images = manifest["images"]
    unless images.is_a?(Array) && !images.empty?
      raise "#{manifest_path}: images must be a non-empty array"
    end

    images.map.with_index do |image, index|
      unless image.is_a?(Hash)
        raise "#{manifest_path}: images[#{index}] must be an object"
      end

      {
        component: component,
        name: require_string(image["name"], "images[#{index}].name", manifest_path),
        source: require_string(image["source"], "images[#{index}].source", manifest_path),
        destination: require_string(
          image["destination"],
          "images[#{index}].destination",
          manifest_path,
        ),
      }
    end
  end

  destinations = {}
  entries.each do |entry|
    destination = entry[:destination]
    raise "duplicate destination image: #{destination}" if destinations[destination]

    destinations[destination] = true
  end

  print JSON.generate(include: entries)
rescue StandardError => error
  warn error.message
  exit 1
end
